extensions/answer_summarizers/models.py

```diff
@@ -38,69 +38,60 @@
-    will return a tuple with all of the list's elements converted to hashable
-    types. If it's a dictionary, it will first convert it to a list of pairs,
-    where the key and value of the pair are converted to hashable types, then
-    it will convert this list as any other list would be converted.
-    """
-    if isinstance(value, list):
-        # Avoid needlessly wrapping a single value in a tuple.
-        if len(value) == 1:
-            return _get_hashable_value(value[0])
-        return tuple([_get_hashable_value(elem) for elem in value])
-    elif isinstance(value, dict):
-        return _get_hashable_value(
-            sorted([
-                (_get_hashable_value(key), _get_hashable_value(val))
-                for (key, val) in value.iteritems()]))
-    else:
-        return value
-
-
-def _count_answers(answer_dicts_list):
-    """Count an input list of answer objects using collections.Counter. This
-    returns a list of pairs with the first element being an answer object and
-    the second being the number of times it shows up in the input list.
-    """
-    hashable_answer_values = [
-        _get_hashable_value(answer_dict['answer'])
-        for answer_dict in answer_dicts_list]
-    answer_frequencies = collections.Counter(hashable_answer_values)
-    return [
-        ([answer_dicts_list[idx]
-          for idx, val in enumerate(hashable_answer_values)
-          if val == hashable_answer][0], frequency)
-        for (hashable_answer, frequency) in answer_frequencies.most_common()]
-
-
-def _calculate_top_answer_frequencies(state_answers_dict, num_results):
+
+class _HashableAnswer(object):
+    """Wraps answer with object that can be placed into sets and dicts."""
+
+    def __init__(self, answer):
+        self.answer = answer
+        self.hashable_answer = utils.get_hashable_value(answer)
+
+    def __hash__(self):
+        return hash(self.hashable_answer)
+
+    def __eq__(self, other):
+        if isinstance(other, _HashableAnswer):
+            return self.hashable_answer == other.hashable_answer
+        return False
+
+
+def _get_top_answers_by_frequency(answers, limit=None):
     """Computes the number of occurrences of each answer, keeping only the top
-    num_results answers, and returns an AnswerFrequencyList.
+    limit answers, and returns an AnswerFrequencyList.
 
     This method is run from within the context of a MapReduce job.
-    """
-    top_answer_counts_as_list_of_pairs = _count_answers(
-        state_answers_dict['submitted_answer_list'])[:num_results]
 
-    answer_frequency_list = stats_domain.AnswerFrequencyList()
-    for item in top_answer_counts_as_list_of_pairs:
-        answer_frequency_list.add_answer(
-            stats_domain.AnswerOccurrence(item[0]['answer'], item[1]))
+    Args:
+        answers: iterable(*). The collection of answers to be tallied.
+        limit: int or None. The maximum numbers of answer to return. When None,
```

Comment:
Btw, small nit -- this should be: The maximum number of answers to return.

Reply:
Done, so close yet so far :')